server/src/main/java/org/elasticsearch/search/aggregations/support/BreakingPriorityQueueWrapper.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.support;
+
+import org.apache.lucene.util.PriorityQueue;
+import org.apache.lucene.util.RamUsageEstimator;
+import org.elasticsearch.common.lease.Releasable;
+
+import java.util.function.Consumer;
+
+/**
+ * A wrapper around Lucene's PriorityQueue which also tracks memory usage via circuit
+ * breakers.
+ *
+ * Lucenes PriorityQueue allocates an array of `size` immediately upon instantiation.
+ * This is potentially very dangerous though if the supplied size is large, and
+ * the caller doesn't correctly account for circuit breaker allocation and de-allocation.
+ *
+ * This class wraps the PQ so that the circuit breaker is correctly called before
+ * and after de-allocation.
+ *
+ * NOTE: this class has _no null checks_ after the queue has been de-allocated (in
+ * order to preserve performance).  Caller must not use the queue after
+ * {@link BreakingPriorityQueueWrapper#close()} is called otherwise an NPE will
+ * be thrown.
+ */
+public abstract class BreakingPriorityQueueWrapper<T> implements Releasable {
```

Comment:
wouldn't it be easier to extend than wrap?

Reply:
Took me a bit to remember why I did it this way.  The constructor of `PriorityQueue` allocates the internal array immediately, so we won't be able to increment the breaker until _after_ the memory has been allocated (and potentially blows up).

Some of the uses of the PQ that's probably ok, but other spots could easily OOM in that call alone I think.

Reply:
What are these spots? The two ones that are covered today seem to me as unlikely to trigger an OOME.

Reply:
The location that kicked off this PR was the SigTerms agg, e.g. here: https://github.com/elastic/elasticsearch/pull/47515/files#diff-a18a5ec0d6ddae35f82515567c98e603R81

If I'm reading everything correctly, we size the PQ to:

```java
final int size = (int) Math.min(bucketOrds.size(), bucketCountThresholds.getShardSize());
```

If the user has selected an abusively large size on the agg, `shardSize` can go to MAX_INT.  Which means the final size is equivalent to the number of bucket ordinals, and on high cardinality fields that can be quite large.  In worst case scenario `size == MAX_INT` too.  Using an 8b pointer in each PQ slot, that comes in at around ~18gb allocated without a breaker check.  

This doesn't account for the actual memory usage of the PQ objects once they are allocated, but the existing max_buckets check should fail out well before the objects themselves can cause memory issues.

This is theoretical, but there were user clusters that OOM'd due to a similar PQ allocation in the sampler agg (albeit used for a different purpose): https://github.com/elastic/elasticsearch/pull/39381

It also just seemed like good practice to account for the memory in the breaker before actually allocating, so that we have a chance to fail before the memory is used.

Reply:
I think you can extend `PriorityQueue` and still call the breaker prior to allocating memory with something like:

```java
public BreakingPriorityQueue(int size, IntUnaryOperator circuitBreaker) {
    super(circuitBreaker.applyAsInt(size));
}
```

Reply:
@jaymode oof, I didn't think about that at all.  :man_facepalming: I'll see swapping this PR over to that, it'd be a lot simpler. Thanks!